Commands/GenerateVisitsDb.php

```diff
@@ -0,0 +1,525 @@
+     * @param int $idSite   Site for which to generate the visits
+     * @param int $limit    Number of visits to generate
+     *
+     * @return array        Statistics array for the data that was generated
+     *                      ['visits' => 0, 'visitActions' => 0, 'actions' => 0, 'conversions' => 0]
+     * @throws \Exception
+     */
+    private function generate(int $time, int $idSite, int $limit): array
+    {
+        $input = $this->getInput();
+        $output = $this->getOutput();
+        $conversionPercent = $input->getOption('conversion-percent');
+        $actionCountMin = $input->getOption('min-actions');
+        $actionCountMax = $input->getOption('max-actions');
+
+        $requestCount = 0;
+        $queryGenerator = new VisitFakeQuery($this->actionsPoolSize);
+
+        if ($this->verbosity == 0) {
+            $output->write(".");
+        }
+
+        $stats = [
+            'visits' => 0,
+            'visitActions' => 0,
+            'actions' => 0,
+            'conversions' => 0
+            ];
+
+        while ($requestCount < $limit || $limit < 0) {
+
+            $this->query(['sql' => 'START TRANSACTION', 'bind' => []]);
+            // Choose a random timestamp within supplied date range
+            $timestampUTC = rand($time, $time + 86000); // no new visits will start for 400 seconds before midnight
+
+            // Create actions for the visit
+            $idactions = [];
+            $actionUrl = '';
+            $actionCount = rand($actionCountMin, $actionCountMax);
+            for ($i = 0; $i != $actionCount; $i++) {
+
+                // Get random action, 50% chance of being new until pool is full, then always an existing action
+                $actionUrl = $queryGenerator->getRandomActionURL();
+
+                // Check if the action exists in the db, create new action if not
+                $findActionQuery = $queryGenerator->getCheckActionExistsQuery($actionUrl);
+                $actionRows = $this->query($findActionQuery, true);
+                if (count($actionRows) === 0) {
+                    if ($this->verbosity === 3) {
+                        $output->writeln("New action, doing insert...");
+                    }
+                    $stats['actions']++;
+
+                    // Insert new action
+                    $insertActionQuery = $queryGenerator->getInsertActionQuery($actionUrl);
+                    $this->query($insertActionQuery, false);
+                    $idactions[] = $this->pdo->lastInsertId();
+                } else {
+                    $idactions[] = $actionRows[0]->idaction;
+                }
+            }
+            $stats['visitActions'] += $actionCount;
+
+            // Get random visitor id (10% chance of a returning visitor id)
+            $idvisitor = $queryGenerator->getVisitor(10);
+            if ($this->verbosity == 3) {
+                $output->writeln("Got idvisitor '".bin2hex($idvisitor)."'");
+            }
```

Comment:
Can we create a local wrapper for `$output->writeln` that would automatically use the verbosity level so we didn't have to wrap it in a condition every time we want to possibly output something for level 3? Then here we would call just e.g. `$this->log(..)` (might need to be a different method name if already taken).

Reply:
Good idea, this makes the code a bit easier to read :+1: